internal/controllers/addon/handle_installed_condition.go

```diff
@@ -54,11 +68,55 @@ func (r *olmReconciler) deleteConfigMapPresent(ctx context.Context, addon *addon
 		Namespace: addonNamespace,
 	}, deleteConfigMap)
 	if err != nil {
-		if errors.IsNotFound(err) {
+		if k8serrors.IsNotFound(err) {
 			return false, nil
 		}
 		return false, err
 	}
 	_, found := deleteConfigMap.Labels[deleteConfigMapLabel]
 	return found, nil
 }
+
+// handleInstallAck handles installation acknowledgement from
+// the addon instance if specified in the addon CR
+func (r *olmReconciler) handleInstallAck(ctx context.Context, addon *addonsv1alpha1.Addon) (requeueResult, error) {
+	installed, err := r.isAddonInstanceInstalled(ctx, addon)
+	if err != nil {
+		// Do not return an error; Wait for the addon's operator to set the status condition
+		if errors.Is(err, errAddonInstanceNotInstalled) {
+			return resultRetry, nil
+		}
+
+		return resultRetry, err
+	}
+
+	if installed {
+		reportInstalledCondition(addon)
+	}
+
+	return resultNil, nil
+}
+
+// isAddonInstanceInstalled returns if the corresponding addon instance has installed=true status condition
+func (r *olmReconciler) isAddonInstanceInstalled(ctx context.Context, addon *addonsv1alpha1.Addon) (bool, error) {
+	addonInstance := &addonsv1alpha1.AddonInstance{}
+	instanceKey := types.NamespacedName{
+		Name:      addonsv1alpha1.DefaultAddonInstanceName,
+		Namespace: GetCommonInstallOptions(addon).Namespace,
+	}
+
+	if err := r.uncachedClient.Get(ctx, instanceKey, addonInstance); err != nil {
+		return false, err
+	}
+
+	if installedCond := meta.FindStatusCondition(
+		addonInstance.Status.Conditions,
+		addonsv1alpha1.AddonInstanceConditionInstalled.String(),
+	); installedCond == nil {
+		// Since the addon instance is awaiting install, we report the addon as unavailable and retry.
+		reportPendingAddonInstanceInstallation(addon)
+		return false, errAddonInstanceNotInstalled
```

Comment:
Why do we need this new error? `isAddonInstanceInstalled` should return true/false

Reply:
The error here is to signal the reconciler to retry. If we return `false, nil` the reconciler will never wait (retry) for the 'Installed' condition to show up in the addon instance.

The logic flow is something like:
In `isAddonInstanceInstalled()` return `errAddonInstanceNotInstalled()` -> In `handleInstallAck` if the error matches then signal a retry (`resultRetry`) instead of returning `resultNil`.

If we just return true/false; incase `installedCond == nil`:
In `isAddonInstanceInstalled()` return false -> In `handleInstallAck()` if `installed==false` returns `resultNil` and the reconciler never retries.

Reply:
You could modify handleInstallAck to return `resultRetry, nil` if installed==false.

Reply:
Makes sense, thanks, had similar changes committed locally, pushed them now